akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala

```diff
@@ -84,6 +90,71 @@ class FusingSpec extends StreamSpec {
       refs.toSet should have size (in.size + 1) // outer/main actor + 1 actor per subflow
     }
 
+    //an UnfoldResourceSource equivalent without an async boundary
+    case class UnfoldResourceNoAsyncBoundry[T, S](create: () => S, readData: (S) => Option[T], close: (S) => Unit)
+        extends GraphStage[SourceShape[T]] {
+      val stage_ = new UnfoldResourceSource(create, readData, close)
+      override def initialAttributes: Attributes = Attributes.none
+      override val shape = stage_.shape
+      def createLogic(inheritedAttributes: Attributes) = stage_.createLogic(inheritedAttributes)
+      def asSource = Source.fromGraph(this)
+    }
+
+    "propagate downstream errors through async boundary" in {
+      val promise = Promise[Done]
+      val slowInitSrc = UnfoldResourceNoAsyncBoundry(
+        () => { Await.result(promise.future, duration.Duration.Inf); () },
+        (_: Unit) => Some(1),
+        (_: Unit) => ()).asSource.watchTermination()(Keep.right).async //commenting this out, makes the test pass
+      val downstream = Flow[Int]
+        .prepend(Source.single(1))
+        .flatMapPrefix(0) {
+          case Nil => throw TE("I hate mondays")
+        }
+        .watchTermination()(Keep.right)
+        .to(Sink.ignore)
+
+      val g = slowInitSrc.toMat(downstream)(Keep.both)
+
+      val (f1, f2) = g.run()
+      f2.failed.futureValue shouldEqual TE("I hate mondays")
+      f1.value should be(empty)
+      //by now downstream managed to fail, hence it already processed the message from Flow.single,
+      //hence we know for sure that all graph stage locics in the downstream interpreter were initialized(=preStart)
+      //hence upstream subscription was initiated.
+      //since we're still blocking upstream's preStart we know for sure it didn't respond to the subscription request
+      //since a blocked actor can process additional messages from its inbox.
+      //so long story short: downstream was able to initialize, subscribe and fail before upstream responded to the subscription request.
+      //prior to akka#29194, this scenario resulted with cancellation signal rather than the expected error signal.
```

Comment:
good with a motivation/explanation for when someone looks at this in three years 👍